coil-core/src/androidMain/kotlin/coil3/util/NetworkObserver.kt

```diff
@@ -50,7 +53,12 @@ private class RealNetworkObserver(
         val request = NetworkRequest.Builder()
             .addCapability(NET_CAPABILITY_INTERNET)
             .build()
-        connectivityManager.registerNetworkCallback(request, networkCallback)
+
+        if (observeDefaultNetwork && Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
```

Comment:
@colinrtwhite I saw in https://github.com/coil-kt/coil/issues/2374 mentioned to include `observeDefaultNetwork` in case someone is using a custom `SocketFactory`. I don't think it would have an impact to the code in this file? They can bind to whatever network they want, but it looks like this code was previously interested in networks with capability `INTERNET`, which could be different than the network used by the app in any case (such as by a custom `SocketFactory`), or is this somehow connected to a passed in `SocketFactory` somewhere I might have missed? 

The reason I am asking, is I am wondering if you even need the `observeDefaultNetwork` capability to begin with. The guidance is to use `registerDefaultNetworkCallback` whenever possible. One of the main reasons, is when you are put on a restricted default network (e.g. a VPN). If you use `registerNetworkCallback`, you'll never receive a callback when that happens as it will filter out all restricted networks even though you have access to them (unless you happen to have the restricted network permission which is OEM controlled and very rare). This is one example, there are others (such as OEM usage of per-app networking features in Android). 

We actually plan to add clear documentation to avoid using `registerNetworkCallback` unless you have a very good reason to do so in the very near future to the Android developer portal to help clear things up.

Thanks.

Reply:
@tech31 Thanks for taking a look! `NetworkObserver` is a weak point in the codebase as I'm not too familiar with the networking APIs. Please let me know if you see other potential improvements (or feel free to submit a PR)!

> in case someone is using a custom SocketFactory

It's possible for a user to provide a custom `OkHttpClient` to `ImageLoader`, which could use a custom `SocketFactory` that performs network requests on a non-default network (as I understand it). It's a bit more clear [on the 2.x branch](https://github.com/coil-kt/coil/blob/2.x/coil-base/src/main/java/coil/ImageLoader.kt#L153) where Coil depends on OkHttp directly. I wanted to err on the side of not changing the default behaviour in the 2.x (stable) branch in case it could be an unexpected change for someone. Though, if you think it's worthwhile we could directly migrate to `registerDefaultNetworkCallback`.

For a bit of context, Coil uses the `NetworkObserver` to determine if the device is "offline". If true, it automatically sends all network requests to the disk cache and fails if the image isn't disk cached.

Reply:
> Reading through your comments it sounds like the ideal would be to use registerDefaultNetworkCallback by default and allow users to provide a custom NetworkObserver if they don't use the default network with their OkHttpClient. This change would probably be good to land in the 3.x releases (currently in alpha). What do you think?

Yeah, that sounds reasonable to me from an implementation perspective. Basically, if you want to allow users to customize the types of networks they care about, you would basically do what you are already doing today by using `registerNetworkCallback`, except that instead of newing up a `NetworkRequest` in `NetworkObserver`, you would allow users to pass in a custom `NetworkRequest` which would then be used. How that would make sense from an implementation detail, e.g. new method on `NetworkObserver` or having a custom observer would be totally up to you.

One note on the above; I'm guessing that 99%+ of your users would be fine with the default network, so you'll probably want to comment any such functionality that users should only use such methods if they really need to, otherwise to rely on the default behavior. What we see on the Android Connectivity side is usage of `registerNetworkCallback` unnecessarily, which can have unintended consequences. We plan to update the guidance on the Android developer portal as well in the near future saying as much ("always use the default network unless you know what you're doing and really need something else" to paraphrase).